ensembleperturbation/perturbation/make_storm_ensemble.py

```diff
@@ -0,0 +1,765 @@
+#! /usr/bin/env python3
+"""
+Script to:
+(1) extract an ATCF best track dataset;
+(2) randomly perturb different parameters (e.g., intensity,
+size, track coordinates) to generate an ensemble; and
+(3) write out each ensemble member to the fort.22 ATCF
+tropical cyclone vortex format file.
+
+Variables that can be perturbed:
+- "max_sustained_wind_speed" (Vmax) is made weaker/stronger
+  based on random gaussian distribution with sigma scaled by
+  historical mean absolute errors. central_pressure (pc)
+  is then changed proportionally based on Holland B
+
+- "radius_of_maximum_winds" (Rmax) is made small/larger
+  based on random number in a range bounded by the 15th and
+  85th percentile CDF of historical forecast errors.
+
+- "along_track" variable is used to offset the coordinate of
+  the tropical cyclone center at each forecast time forward or
+  backward along the given track based on a random gaussian
+  distribution with sigma scaled by historical mean absolute
+  errors.
+
+- "cross_track" variable is used to offset the coordinate of
+  the tropical cyclone center at each forecast time a certain
+  perpendicular distance from the given track based on a
+  random gaussian distribution with sigma scaled by historical
+  mean absolute errors.
+
+By William Pringle, Argonne National Laboratory, Mar-May 2021
+   Zach Burnett, NOS/NOAA
+   Saeed Moghimi, NOS/NOAA
+"""
+
+from abc import ABC
+from argparse import ArgumentParser
+from copy import deepcopy
+from datetime import datetime, timedelta
+from enum import Enum
+from math import exp, inf, sqrt
+from os import PathLike
+from pathlib import Path
+from random import gauss, random
+
+from adcircpy.forcing.winds.best_track import BestTrackForcing
+from dateutil.parser import parse as parse_date
+from numpy import append, floor, insert, interp, sign, transpose
+from pandas import DataFrame
+from pyproj import Proj
+from shapely.geometry import LineString
+
+AIR_DENSITY = 1.15  # [kg/m3]
+
+UNIT_CONVERSIONS = {
+    'knot': {'meter_per_second': 0.514444444},
+    'millibar': {'pascal': 100},
+    'pascal': {'millibar': 0.01},
+    'nautical_mile': {'statute_mile': 1.150781, 'meter': 1852},
+    'statute_mile': {'nautical_mile': 0.868976},
+}
+
+E1 = exp(1.0)  # e
+
+# Index of absolute errors (forecast times [hrs)]
+ERROR_INDICES_NO_60H = [0, 12, 24, 36, 48, 72, 96, 120]  # no 60-hr data
+ERROR_INDICES_60HR = [0, 12, 24, 36, 48, 60, 72, 96, 120]  # has 60-hr data (for Rmax)
+
+
+class RandomType(Enum):
+    GAUSSIAN = 'gaussian'
+    LINEAR = 'linear'
+
+
+class BestTrackPerturbedVariable(ABC):
+    name: str
+    random_type: RandomType = None
+    lower_bound: float = None
+    upper_bound: float = None
+    historical_forecast_errors: {str: float} = None
+    default: float = None  # [mbar]
```

Comment:
implement new variables by extending this abstract class (see `CentralPressure`, below, for example)